pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,546 @@
+			return
+		}
+	}
+	c.queue.Add(key{
+		kind: "Gateway",
+		name: gw.Name,
+	})
+}
+
+func (c *MCRouteController) enqueueClusterInfoImport(obj interface{}) {
+	ciImp, isciImp := obj.(*mcv1alpha1.ClusterInfoImport)
+	if !isciImp {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("received unexpected object: %v", obj)
+			return
+		}
+		ciImp, ok = deletedState.Obj.(*mcv1alpha1.ClusterInfoImport)
+		if !ok {
+			klog.Errorf("DeletedFinalStateUnknown contains non-ClusterInfoImport object: %v", deletedState.Obj)
+			return
+		}
+	}
+
+	c.queue.Add(key{
+		kind: "ClusterInfoImport",
+		name: ciImp.Name,
+	})
+}
+
+// Run will create defaultWorkers workers (go routines) which will process
+// the Gateway events from the workqueue.
+func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+	cacheSyncs := []cache.InformerSynced{c.gwListerSynced, c.ciImportListerSynced}
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem
+// function in order to read and process a message on the workqueue.
+func (c *MCRouteController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MCRouteController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if k, ok := obj.(key); !ok {
+		c.queue.Forget(obj)
+		klog.InfoS("Expected type 'key' in work queue but got object", "object", obj)
+		return true
+	} else if err := c.syncMCFlows(k); err == nil {
+		c.queue.Forget(k)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCRouteController) syncMCFlows(k key) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing Gateway Node flows for Multi-cluster", "type", k.kind,
+			"name", k.name, "time", time.Since(startTime))
+	}()
+	switch k.kind {
+	case "Gateway":
+		klog.V(2).InfoS("Processing for Gateway", "gateway", k.name)
+		return c.handleGatewayEvents(k)
+	case "ClusterInfoImport":
+		klog.V(2).InfoS("Processing for ClusterInfoImport", "clusterinfoimport", k.name)
+		return c.handleCIImportEvents(k)
+	}
+	return nil
+}
+
+func (c *MCRouteController) handleGatewayEvents(k key) error {
+	var gwIsActive, iAmActiveGW, noActiveGateway bool
+	activeGW, iamGW, getGWErr := c.checkActiveGateway()
+	klog.V(2).InfoS("Got the active Gateway", "gateway", klog.KObj(activeGW))
+	if getGWErr != nil {
+		if strings.Contains(getGWErr.Error(), "no Gateway found") {
```

Comment:
This is not right. I think either use nil to express not found, or add a boolean. And this is not an error.

Reply:
Done, this part has been refined into multiple small methods.